src/game_state/Octree.cpp

```diff
@@ -0,0 +1,185 @@
+#include "Octree.h"
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <cstdlib>
+
+Octree::Octree(
+    std::shared_ptr<std::vector<std::shared_ptr<GameObject>>> objects)
+    : root(constructOctree(objects.get())) {
+  this->objects = objects;
+}
+
+Octree::~Octree() {
+  delete root;
+}
+
+Node* Octree::GetRoot() {
+  return root;
+}
+
+std::shared_ptr<std::vector<std::shared_ptr<GameObject>>> Octree::getObjects() {
+  return objects;
+}
+
+Node* Octree::constructOctree(std::vector<std::shared_ptr<GameObject>>* objs) {
+  // leaf
+  if (objs->size() <= OBJS_IN_LEAF) {
+    glm::vec3 min, max;
+    if (objs->empty()) {
+      std::cerr << "Catastrophic system failure. Unrecoverable" << std::endl;
+      exit(EXIT_FAILURE);
+    }
+    return new Node(objs, AxisAlignedBox(getMin(objs), getMax(objs)),
+                    new std::vector<Node*>());
```

Comment:
Oh wow really? Maybe we should reconsider shared_ptr in some other places. How big was the performance hit? Like 10 fps or something?